### PR TITLE
Added `where` kwarg to `astropy.units.Quantity`'s reduction-like methods

### LINKED ISSUE #12890: The `where` keyword argument of `np.mean` is not supported for `astropy.units.Quantity` instances.
### Description

Apologies if there is a duplicate somewhere, I scoured all the issues for this problem and I couldn't find it mentioned yet.

The `where` keyword argument was added to `np.mean` and all the other `np.reduce`-based functions in version 1.20.0 of numpy,
but it doesn't seem to work yet with `astopy.units.Quantity`.

Does anyone know if there is some limitation in `astropy.units.Quantity` that is preventing this feature from being implemented?

If not, I could put some time towards updating `astropy.units.Quantity` to support this feature.

### PR DESCRIPTION
**Pull request by byrdie**
Added `where` keyword (and `initial` keyword where appropriate) to the following methods:
 - `astropy.units.Quantity.var()`
 - `astropy.units.Quantity.std()`
 - `astropy.units.Quantity.mean()`
 - `astropy.units.Quantity.nansum()`

### Description

This pull request is to address issues with the `where` keyword added to `numpy.mean()` in numpy version 1.20.0.

Fixes #12890

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Do the proposed changes actually accomplish desired goals?
- [x] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [x] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [x] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [x] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [ ] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label.
- [x] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [x] Is this a big PR that makes a "What's new?" entry worthwhile and if so, is (1) a "what's new" entry included in this PR and (2) the "whatsnew-needed" label applied?
- [x] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [x] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

**Review by mhvk — COMMENTED** [review]
The changes look good in principle, but you have to be careful that numpy < 1.20 do not understand the `where` argument, i.e., you need something like
```
# on top
from astropy.utils.compat import NUMPY_LT_1_20

# for each method
if NUMPY_LT_1_20:
    def var(...):
        # old version
else:
    # new version
```

Unfortunately, it looks like most of the failures are actually because the `Masked` class also does not have `where` and `initial` -- see `astropy/utils/masked/core.py`. If you are interested in addressing that too, you would need to be careful that the arguments are interpreted correctly (i.e., that `where` is combined properly with the mask).

Alternatively, to keep this PR smaller, it may be possible to use defaults of `np._NoValue`, which I think get stripped correctly by the numpy functions called by `_wrap_function`. In that case, we would just need an issue to remind us to do the masked versions too.

Also, could you ensure the signature is as close as possible to the numpy one? I.e., insert an `, *, where=...` to ensure the new arguments are keyword-only.

File: astropy/utils/masked/core.py
Comment (byrdie) [inline astropy/utils/masked/core.py:1078]: @mhvk, do you know why the mask wasn't used on this line before? To my naive understanding, it seems like the mask should be used here.

**Comment by byrdie** [comment]
I've also modified the methods within `Quantity` to change the method signature based on the numpy version per your suggestion. 

I haven't applied the same logic to `MaskedNDArray`'s methods yet, is it surprising that the tests are passing?

File: astropy/utils/masked/core.py
Comment (byrdie) [reply astropy/utils/masked/core.py:1078]: Ok, that makes sense to me now, I needed to reset my brain to think in terms of masked arrays.

File: astropy/utils/masked/core.py
Comment (byrdie) [inline astropy/utils/masked/core.py:1080]: @mhvk, our conversation above makes me wonder why the mask had to be passed into `sum()` explicitly here.
Isn't `x` a `MaskedNDArray`, making the use of the `where` keyword argument redundant in the original implementation?

If we decide it is redundant, I need to change the patch to be `where=where` instead of `where=where_final`

Comment (mhvk) [reply astropy/utils/masked/core.py:1080]: By default, the mask is propagated in `.sum()` so here it has to be passed explicitly that we want to ignore masked elements. This is one of the differences with `np.ma.MaskedArray`; for the rationale, see https://docs.astropy.org/en/latest/utils/masked/index.html#utils-masked-vs-numpy-maskedarray

p.s. Hope to be able to review soon, but things are a bit hectic right now.

Comment (byrdie) [reply astropy/utils/masked/core.py:1080]: Thanks for showing me that documentation, I was not aware of that difference.

No worries on the review, and thanks for the feedback that you've already provided.

**Comment by nstarman** [comment]
@byrdie, thanks for the PR.
I've re-requested a review from @mhvk.
Just as a note, we prefer rebasing branches onto main, not merging main into a branch, so if you could roll back the last commit and rebase instead, that would be great.

**Comment by byrdie** [comment]
@nstarman, thanks for pointing out the issue with the merge commit.

Unfortunately there's another merge commit, 6a3a1872 that was made before a bunch of commits to the feature.
I've attempted to reset my branch to the commit before the first merge commit, and then I cherry-picked my commits and force-pushed to my branch. Hopefully that was the best way of resolving the issue.

**Comment by nstarman** [comment]
@byrdie. This looks really good. Thanks for the PR! Can you add some tests for these new features? Then this will probably be ready to merge. Thanks!

**Comment by mhvk** [comment]
I looked only briefly, but this indeed looks good, modulo a few tests -- Don't worry about being completely exhaustive! But enough that we can check a few explicit cases make sense (and that they fail on older numpy as appropriate).

File: astropy/units/tests/test_quantity_array_methods.py
Comment (nstarman) [inline astropy/units/tests/test_quantity_array_methods.py]: Feel free to ignore, but the test assertion is immediately obvious if it's Gaussian summation.

```suggestion
        q1 = np.array([1., 2., 3., 4., 5., 6., 7.]) * u.m
        assert_array_equal(np.mean(q1, where=q1 < 7), 3.5 * u.m)
```

**Review by nstarman — APPROVED** [review]
LGTM. Thanks @byrdie!
Let's wait for @mhvk to give final approval, as this touches Masked, which is out of my wheelhouse.

File: astropy/units/tests/test_quantity_array_methods.py
Comment (byrdie) [reply astropy/units/tests/test_quantity_array_methods.py]: I agree that is more clear.

The sequence I used was used in the two previous tests concerning `np.mean()`, `test_mean()` and `test_mean_inplace()`, and I thought I would keep things consistent. What do you think is most important?

Comment (nstarman) [reply astropy/units/tests/test_quantity_array_methods.py]: For me, clarity. I'm even tempted to change the other tests for consistency 😆

Comment (byrdie) [reply astropy/units/tests/test_quantity_array_methods.py]: Do you think changing the other tests is within the scope of this PR?

**Comment by byrdie** [comment]
@nstarman and @mhvk, even though not addressed by changes in this PR, I've added tests for using the `where` keyword argument in `numpy.sum()`, `numpy.min()`, and `numpy.max()`. Do you think that is appropriate?

**Review by nstarman — CHANGES_REQUESTED** [review]
@byrdie CI failure appears real.

**Comment by byrdie** [comment]
@nstarman, I've finished adding tests to this PR. Can you take a look when convenient?

I know some of the checks are failing, do you know if that's my fault, or are those checks expected to fail right now?

**Comment by pllim** [comment]
p.s. Please also squash your commits.

File: astropy/units/tests/test_quantity_array_methods.py
Comment (nstarman) [reply astropy/units/tests/test_quantity_array_methods.py]: I think best to leave it.

**Review by nstarman — APPROVED** [review]
@byrdie. This looks really good to me. Thanks for the contribution! Let's have @mhvk weigh in on the Masked stuff before merging.

**Comment by nstarman** [comment]
Hm. Looks like you did a merge commit? Please rebase instead.

**Comment by byrdie** [comment]
@nstarman, I accidentally rebased onto origin/main instead of astropy/main. Trying to sort out the issue right now.

File: docs/changes/units/12891.feature.rst
Comment (byrdie) [inline docs/changes/units/12891.feature.rst:4]: Are we allowed to use unordered lists in the towncrier log entry?

File: docs/changes/utils/12891.feature.rst
Comment (byrdie) [inline docs/changes/utils/12891.feature.rst:5]: Are we allowed to use the same filename, `12891.feature.rst`, as long as it's in a different path?

Comment (pllim) [reply docs/changes/utils/12891.feature.rst:5]: I think so... but why can't this be one sentence like this?

*Added there `where` keyword argument to X, Y, ..., and Z*

Comment (byrdie) [reply docs/changes/utils/12891.feature.rst:5]: Ok, so the preferred way to do this is two files, with one sentence in each?

Comment (pllim) [reply docs/changes/utils/12891.feature.rst:5]: I think so but @saimn can confirm.

**Comment by nstarman** [comment]
@mhvk, aside from the file conflict, is this PR good to go into v5.1 or would you rather it be in v5.2?

**Review by mhvk — APPROVED** [review]
This looks good!

**Comment by pllim** [comment]
This has 2 approvals and failure is unrelated. So merging. Thanks!

**Comment by pllim** [comment]
Belatedly -- @mhvk , this does not need a What's New?

**Comment by mhvk** [comment]
Don't think so - in a way it was a bug that we did not support the same arguments as the numpy equivalents.

**Pull request merged**
